src/workload_api/client.rs

```diff
@@ -152,84 +203,117 @@ impl WorkloadApiClient {
     ///
     /// The function returns a variant of [`ClientError`] if environment variable is not set or if
     /// the provided socket path is not valid.
-    pub fn default() -> Result<Self, ClientError> {
+    pub async fn default() -> Result<Self, ClientError> {
         let socket_path = match get_default_socket_path() {
             None => return Err(ClientError::MissingEndpointSocketPath),
             Some(s) => s,
         };
-        Self::new(socket_path.as_str())
+        Self::new_from_path(socket_path.as_str()).await
+    }
+
+    /// Constructs a new `WorkloadApiClient` using the provided Tonic transport channel.
+    ///
+    /// # Arguments
+    ///
+    /// * `conn`: A `tonic::transport::Channel` used for gRPC communication.
+    ///
+    /// # Returns
+    ///
+    /// A `Result` containing a `WorkloadApiClient` if successful, or a `ClientError` if an error occurs.
+    pub fn new(conn: tonic::transport::Channel) -> Result<Self, ClientError> {
+        Ok(WorkloadApiClient {
+            client: SpiffeWorkloadApiClient::with_interceptor(conn, MetadataAdder {}),
+        })
     }
 
-    /// Fetches the default [`X509Svid`], i.e. the first in the list returned by the Workload API.
+    /// Fetches a single X509 SPIFFE Verifiable Identity Document (SVID).
+    ///
+    /// This method connects to the SPIFFE Workload API and returns the first X509 SVID in the response.
+    ///
+    /// # Returns
+    ///
+    /// On success, it returns a valid [`X509Svid`] which represents the parsed SVID.
+    /// If the fetch operation or the parsing fails, it returns a [`ClientError`].
     ///
     /// # Errors
     ///
-    /// The function returns a variant of [`ClientError`] if there is en error connecting to the Workload API or
-    /// there is a problem processing the response.
-    pub fn fetch_x509_svid(&self) -> Result<X509Svid, ClientError> {
-        let request = X509SVIDRequest::new();
+    /// Returns [`ClientError`] if the gRPC call fails or if the SVID could not be parsed from the gRPC response.
+    pub async fn fetch_x509_svid(&mut self) -> Result<X509Svid, ClientError> {
+        let request = X509svidRequest::default();
 
-        let mut response = self
-            .client
-            .fetch_x509_svid_opt(&request, WorkloadApiClient::call_options()?)?;
+        let grpc_stream_response: tonic::Response<tonic::Streaming<X509svidResponse>> =
+            self.client.fetch_x509svid(request).await?;
 
-        let item = block_on(response.next());
-        let x509_svid = WorkloadApiClient::process_x509_svid(item)?;
-        Ok(x509_svid)
+        let response = match grpc_stream_response.into_inner().message().await? {
+            Some(response_value) => response_value,
+            None => return Err(ClientError::EmptyResponse),
+        };
```

Comment:
Nit: can replace these blocks with `ok_or` for example: 
```suggestion
let response = grpc_stream_response.into_inner().message().await?.ok_or(ClientError::EmptyResponse)?;
```

Reply:
Thank you for the suggestion! I've made the suggested change. I replaced the existing blocks with `ok_or` as it does make the code cleaner and more idiomatic.